lib/rest-simple-interface.js

```diff
@@ -0,0 +1,298 @@
+
+		const getQueryParam = (name) => {
+			let str = req.query && req.query[name];
+			if (str) {
+				try {
+					return JSON.parse(str);
+				} catch (err) {
+					return str;
+				}
+			} else {
+				return undefined;
+			}
+		};
+
+		const findObjectOperationMethod = (methodSuffix) => {
+			for (let i = pathComponents.length - 1; i >= 1; i--) {
+				let methodName = pathComponents.slice(0, i).join('.') + '.' + methodSuffix;
+				if (this.methods[methodName] && this.methods[methodName].options.model) {
+					// Make sure the number of path keys and model keys matches
+					let modelKeys = this.methods[methodName].options.model.getKeys();
+					if (modelKeys.length === pathComponents.length - i) {
+						// Construct a keys object
+						let keys = {};
+						for (let j = 0; j < modelKeys.length; j++) {
+							keys[modelKeys[j]] = pathComponents[i + j];
+						}
+						// Return the results
+						return {
+							method: methodName,
+							keys
+						};
+					}
+				}
+			}
+			return null;
+		};
+
+
+		if (req.method === 'POST') {
+			// As a POST, this is either an object creation, or a method call.
+			// First check if there's a corresponding method call.
+			let methodName = pathComponents.join('.');
+			if (this.methods[methodName]) {
+				// A method exists by that name, so treat this as a method call
+				let methodParams = (req.body && _.isPlainObject(req.body)) ? req.body : {};
+				let methodMiddleware = this.methods[methodName].middleware;
+				this.handleAPICall(req, res, this.methods[methodName].options, methodParams, ...methodMiddleware);
+				return;
+			}
+
+			// Check if there's a corresponding put method for object creation (handled same as object replacement)
+			methodName = pathComponents.join('.') + '.put';
+			if (this.methods[methodName]) {
+				// A corresponding put method exists, so call it with the given data
+				let methodParams = {
+					data: body
+				};
+				let methodMiddleware = this.methods[methodName].middleware;
+				this.handleAPICall(req, res, this.methods[methodName].options, methodParams, ...methodMiddleware);
+				return;
+			}
+		}
+
+		if (req.method === 'GET') {
+			// This is either a get on an individual object, or a get on an object root (a list)
+			// First check if it fits the pattern for a get on a single object
+			// Because objects can have multiple keys, multiple route prefixes need to be checked
+			let so = findObjectOperationMethod('get');
+			if (so) {
+				// Run get on single object
+				let methodParams = {
+					keys: so.keys,
+					fields: getQueryParam('fields')
+				};
+				let methodMiddleware = this.methods[so.method].middleware;
+				this.handleAPICall(req, res, this.methods[so.method].options, methodParams, ...methodMiddleware);
+				return;
+			}
+
+			// Check for GET on object root (list/query)
+			let methodName = pathComponents.join('.') + '.query';
+			if (this.methods[methodName]) {
+				let methodParams = {
+					query: getQueryParam('query') || {},
```

Comment:
So query objects are passed as serialized JSON on the query string? I guess there's no better option but yuck

Reply:
Any consideration to how query presets should be presented in REST? Currently it would be a POST to the preset method, which probably looks strange and inconsistent to the RESTing eye, especially since query presets will be the more common way to interact with the API. This is even further complicated by the fact that query presets also support aggregation, which I think belongs as a POST anyway.

Reply:
Query objects are either passed in the query string (only option if doing a GET) or in the body as a POST request.

For preset queries, you really only have the same two options: either implement it as a separate POSTable method, or add the ability to provide a 'preset' parameter to query (although we've already discussed the problems with this approach).

REST just sucks for stuff like this.